File: thinking-categorically/state-monad/06-configuration-management-pattern/configuration-management-pattern.ts
class State<S, A> {
    constructor(private runState: (state: S) => [A, S]) {}

    run(initialState: S): [A, S] {
        return this.runState(initialState);
    }

    map<B>(f: (a: A) => B): State<S, B> {
        return new State(s => {
            const [a, newS] = this.runState(s);
            return [f(a), newS];
        });
    }

    flatMap<B>(f: (a: A) => State<S, B>): State<S, B> {
        return new State(s => {
            const [a, newS] = this.runState(s);
            return f(a).run(newS);
        });
    }

    then<B>(next: State<S, B>): State<S, B> {
        return this.flatMap(_ => next);
    }

    static pure<S, A>(value: A): State<S, A> {
        return new State(s => [value, s]);
    }

    static get<S>(): State<S, S> {
        return new State(s => [s, s]);
    }

    static put<S>(newState: S): State<S, void> {
        return new State(_ => [undefined, newState]);
    }
}

// Application configuration interface
interface AppConfig {
    apiUrl: string;
    timeout: number;
    features: Set<string>;
}

// Enable a feature if not already enabled
function enableFeature(feature: string): State<AppConfig, boolean> {
    return State.get<AppConfig>()
        .flatMap(config => {
            if (config.features.has(feature)) {
                return State.pure(false); // Already enabled
            } else {
                const newFeatures = new Set(config.features);
                newFeatures.add(feature);
                return State.put<AppConfig>({
                    ...config,
                    features: newFeatures
                }).map(() => true);
            }
        });
}

// Configure multiple features
function configureApp(): State<AppConfig, string> {
    return enableFeature("darkMode")
        .flatMap(darkMode =>
            enableFeature("notifications")
                .flatMap(notifications =>
                    enableFeature("analytics")
                        .map(analytics =>
                            `Configured: darkMode=${darkMode}, notifications=${notifications}, analytics=${analytics}`
                        )
                )
        );
}

// Usage example
const initialConfig: AppConfig = {
    apiUrl: "https://api.example.com",
    timeout: 5000,
    features: new Set<string>()
};

const [result, finalConfig] = configureApp().run(initialConfig);
console.log(result);
console.log("Final config:", {
    ...finalConfig,
    features: Array.from(finalConfig.features)
});
// Output shows which features were newly enabled

File: thinking-categorically/state-monad/06-configuration-management-pattern/configuration-management-pattern.cs
using System;
using System.Collections.Generic;
using System.Linq;

// Unit type for void-like operations
public struct Unit
{
    public static readonly Unit Default = new Unit();
}

public class State<S, A>
{
    private readonly Func<S, (A Result, S NewState)> computation;

    public State(Func<S, (A, S)> computation)
    {
        this.computation = computation;
    }

    public (A Result, S NewState) Run(S initialState)
    {
        return computation(initialState);
    }

    public State<S, B> Map<B>(Func<A, B> f)
    {
        return new State<S, B>(s =>
        {
            var (result, newState) = computation(s);
            return (f(result), newState);
        });
    }

    public State<S, B> FlatMap<B>(Func<A, State<S, B>> f)
    {
        return new State<S, B>(s =>
        {
            var (result, newState) = computation(s);
            return f(result).Run(newState);
        });
    }

    // LINQ Support
    public State<S, B> SelectMany<B>(Func<A, State<S, B>> f)
    {
        return FlatMap(f);
    }

    public State<S, C> SelectMany<B, C>(Func<A, State<S, B>> f, Func<A, B, C> projection)
    {
        return FlatMap(a => f(a).Map(b => projection(a, b)));
    }

    public State<S, B> Select<B>(Func<A, B> f)
    {
        return Map(f);
    }

    public static State<S, A> Pure(A value)
    {
        return new State<S, A>(s => (value, s));
    }

    public static State<S, S> Get()
    {
        return new State<S, S>(s => (s, s));
    }

    public static State<S, Unit> Put(S newState)
    {
        return new State<S, Unit>(_ => (Unit.Default, newState));
    }
}

// Application configuration
public class AppConfig
{
    public string ApiUrl { get; set; }
    public int Timeout { get; set; }
    public HashSet<string> Features { get; set; }

    public AppConfig(string apiUrl = "", int timeout = 0, HashSet<string>? features = null)
    {
        ApiUrl = apiUrl;
        Timeout = timeout;
        Features = features ?? new HashSet<string>();
    }

    public override string ToString()
    {
        return $"AppConfig(ApiUrl: \"{ApiUrl}\", Timeout: {Timeout}, Features: [{string.Join(", ", Features)}])";
    }
}

public static class Program
{
    // Enable a feature if not already enabled
    public static State<AppConfig, bool> EnableFeature(string feature)
    {
        return from config in State<AppConfig, AppConfig>.Get()
               let alreadyEnabled = config.Features.Contains(feature)
               from _ in alreadyEnabled ? State<AppConfig, Unit>.Put(config)
                                        : State<AppConfig, Unit>.Put(new AppConfig(config.ApiUrl, config.Timeout,
                                            new HashSet<string>(config.Features) { feature }))
               select !alreadyEnabled;
    }

    // Configure multiple features
    public static State<AppConfig, string> ConfigureApp()
    {
        return from darkMode in EnableFeature("darkMode")
               from notifications in EnableFeature("notifications")
               from analytics in EnableFeature("analytics")
               select $"Configured: darkMode={darkMode}, notifications={notifications}, analytics={analytics}";
    }

    // Usage example
    public static void Main()
    {
        var initialConfig = new AppConfig("https://api.example.com", 5000);
        var (result, finalConfig) = ConfigureApp().Run(initialConfig);

        Console.WriteLine(result);
        Console.WriteLine($"Final config: {finalConfig}");
        // Output shows which features were newly enabled
    }
}
